### ISSUE TITLE
Distinguer responsable d'organisation et administrateur

### ISSUE THREAD

**Issue by florimondmanca** [etalab/catalogage-donnees#288]
## Description du problème

Actuellement, on a 2 rôles : `USER` et `ADMIN`

Cela permet déjà de faire une distinction en permettant par exemple aux seuls ADMIN de supprimer une fiche de données.

Lors de l'atelier du 10/06, le besoin d'un 3e rôle s'est fait sentir, dans l'objectif de permettre à des "super-utilisateurs" de parcourir, accéder et éventuellement modifier l'ensemble des catalogues (équipe technique, DINUM).

Mais un simple enum comme `User.role` ne peut plus suffire. En effet, on n'est pas MANAGER _en soi_, mais _pour une organisation_. Alors comment faire ?

## Solution envisagée

Il me semble qu'il faut intégrer cette notion de "responsable d'organisation" au modèle de données lui-même.

D'où :

* Convertir la notion d'administrateur en un simple booléen : `User.is_admin`. Si celui-ci est activé, on a "tous les droits".
* Créer une relation `User (0,1)---(0,n) Organization` représentant le fait qu'un utilisateur est membre d'une organisation. NB : un utilisateur peut n'être lié à aucune organisation, juste au moment de sa création.
* Créer une relation 1-1 entre un `User` particulier et un `OrganizationOwner`, représentant le fait d'être "LE responsable" d'une organisation.

Inspiration : https://pypi.org/project/django-organizations/

Modèle physique :

```yaml
User:
  id: ID (pk)
  is_admin: bool
  organization_id: ID | None

OrganizationOwner:
  id: ID (pk)
  user_id: ID
  organization_id: ID

Organization:
  id: ID (pk)
  created_at: datetime
  members: relationship("User")
  owner: relationship("OrganizationOwner")
```

Pseudo-code applicatif :

```python
# Création d'une organisation
multi = execute(CreateOrganization(name="Multi"))

# Affectation d'un 1er utilisateur à l'organisation : il devient automatiquement responsable
# de l'organisation (= devient son `owner`)
johan = execute(CreateUser(..., organization_id=multi.id))
johan in multi.members  # True
multi.owner.user == johan  # True

# Ajout d'un 2e utilisateur : c'est un simple membre
romain = execute(CreateUser(..., organization_id=multi.id))
romain in multi.members  # True
multi.owner.user  # Toujours `johan`
```

Permissions :

```python
# L'utilisateur :user_id peut-il contribuer au catalogue de l'organisation :organization_id?
can_contribute = user.organization_id == organization_id

# L'utilisateur :user_id peut-il retirer des jeux de données du catalogue de l'organisation :organization_id?
can_delete_datasets = user.organization_id == organization.id AND user.id == organization.owner_id
```

Autres références :

* https://learndjango.com/tutorials/django-best-practices-user-permissions
* https://forum.djangoproject.com/t/how-are-you-handling-user-permissions-in-more-complex-projects/2430/10

**Comment by florimondmanca** [comment]
@johanricher Je me demande ce qu'il faut faire de ce ticket ?

Dans le cadre de #388 je me rends compte qu'il y a toujours de la place pour une notion de "responsable d'organisation" (ex : Roselyne Aliacar) et "(super)admin" (nous).

Les cas d'usage concrets sont dans la description du ticket, je pense en particulier à ceux-ci :

* Un super-admin peut modifier n'importe quel jeu de données existant. Un respo d'organisation n'a pas de droit particulier en édition, il ne peut modifier qu'au sein du catalogue de son orga comme n'importe quel USER.
* Un super-admin peut supprimer n'importe quel jeu de données existant. Un respo d'organisation ne doit pouvoir le faire qu'au sein de son orga.

Ça demande un peu de travail.

Si on pense ne pas avoir le temps (j'ai du mal à estimer, il y a encore d'autres peaufinages à faire), je pense qu'il faudrait renommer le rôle ADMIN par exemple en ORG_OWNER. Dès lors nous (équipe de dev rattachée à l'orga "legacy") n'aurions pas de droit en écriture côté UI vis-à-vis des catalogues (on peut tjr faire ce qu'on veut dans la DB), seulement en lecture comme n'importe quel USER.

**Comment by johanricher** [comment]
Je suis d'accord que ce ticket n'a pas de finalité en soi, c'est à dire qu'il ne justifie pas à lui seul de réaliser des tâches spécifiques. En revanche il sert à avoir une vision d'ensemble plus cohérente sur la façon d'implémenter des tâches qui sont spécifiées sur d'autres tickets.

Comme toujours la logique est de voir au cas par cas ce qui doit être implémenté pour valider les critères d'acceptation. **Ma doctrine est la suivante : s'il y a des éléments nouveaux ou des détails qui apparaissent pour implémenter un ticket, cela veut dire qu'il faut les extraire du périmètre et les implémenter plus tard.**

Dans l'exemple que tu cites (#288) : l'idée c'est que la restriction doit s'appliquer à tout le monde (`user` et `admin) sauf aux `super_admin`. Puisque c'est une forme d'exception qui pose des questions, c'est à implémenter séparément et ce n'est pas un must have pour la livraison.

Je vais mettre à jour #288 et créer un ou plusieurs tickets dédiés à l'implé côté super_admin (qui ne seront pas dans le périmètre de la livraison).
